### PR TITLE
[Docs] Client side caching topic

### LINKED ISSUE #1137: Document PWA Client-side Caching
The GraphQL team needs documentation regarding how PWA implements client-side caching. This PWA ticket is related to the following GraphQL ticket: https://github.com/magento/graphql-ce/issues/18

- Service workers as a network cache on the client
- How Apollo maintains a cache of query results client-side
- etc.

### PR DESCRIPTION
**Pull request by jcalcaben**
## Description

Create documentation describing the client-side caching strategies used in Venia.

## Related Issue

Closes #1137 

## How Have YOU Tested this?

Ran the linter in the `pwa-devdocs` project:
```
npm run lint -- src/technologies/basic-concepts/client-side-caching/index.md
```
## Checklist:
- [x] I have updated the documentation accordingly, if necessary.
- [ ] I have added tests to cover my changes, if necessary.

### CODE REVIEW

File: pwa-devdocs/src/technologies/basic-concepts/client-side-caching/index.md [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+Client-to-server communication is slow and expensive.
+Performance is an important feature for any Progressive Web Application(PWA), so
+requests to the server should be minimized.
+
+Offline mode is also a required feature for a PWA.
+In offline mode, the application must be able to serve pages that have been recently viewed.
+
+These features are implemented with the help of a client-side cache.
+This local cache stores data from previously fetched resources.
+Future requests to that resource can use the cache in place of a server request to boost performance.
```
Comment (jimbo) [inline pwa-devdocs/src/technologies/basic-concepts/client-side-caching/index.md]:
```suggestion
Once a resource has been cached, the service worker may consult the cache on subsequent requests for that resource to boost performance.
```

File: pwa-devdocs/src/technologies/basic-concepts/client-side-caching/index.md [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+Client-to-server communication is slow and expensive.
+Performance is an important feature for any Progressive Web Application(PWA), so
+requests to the server should be minimized.
+
+Offline mode is also a required feature for a PWA.
+In offline mode, the application must be able to serve pages that have been recently viewed.
+
+These features are implemented with the help of a client-side cache.
+This local cache stores data from previously fetched resources.
```
Comment (jimbo) [inline pwa-devdocs/src/technologies/basic-concepts/client-side-caching/index.md]:
```suggestion
This local cache stores data from resources as they are fetched.
```

File: pwa-devdocs/src/technologies/basic-concepts/client-side-caching/index.md [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+A [service worker][] is a JavaScript file that runs in a separate thread from the main execution thread in a web application.
+Service workers can intercept network requests and fetch cached data or store results from a network request into the cache.
+
+### Venia service worker
+
+Venia's service worker behavior is defined in the [`src/sw.js`][] file using Google's [Workbox][] library.
+
+You do not need to use Workbox to define service worker behavior, but
+Workbox makes this task easier by removing boilerplate code that is always used when working with service workers.
```
Comment (jimbo) [inline pwa-devdocs/src/technologies/basic-concepts/client-side-caching/index.md]: Helpful clarification. 👍

File: pwa-devdocs/src/technologies/basic-concepts/client-side-caching/index.md [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+
+Venia uses the following [caching strategies][] with its service worker:
+
+#### [Stale-while-revalidate][]
+
+The stale-while-revalidate strategy tell the service worker to use a cached response if it exists.
```
Comment (jimbo) [inline pwa-devdocs/src/technologies/basic-concepts/client-side-caching/index.md]:
```suggestion
The stale-while-revalidate strategy tells the service worker to use a cached response if it exists.
```

**Comment by jcalcaben** [comment]
@jimbo Thanks for the review. I made the suggested changes.

File: pwa-devdocs/src/technologies/basic-concepts/client-side-caching/index.md [hunk 5]
```diff
@@ -0,0 +1,89 @@
+---
+title: Client-side caching
+---
+
+Client-server communication is slow and expensive.
+Performance is an important feature for any Progressive Web Application(PWA), so
```
Comment (cherdman) [inline pwa-devdocs/src/technologies/basic-concepts/client-side-caching/index.md]:
```suggestion
Performance is an important feature for any Progressive Web Application (PWA), so
```

File: pwa-devdocs/src/technologies/basic-concepts/client-side-caching/index.md [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+The stale-while-revalidate strategy tells the service worker to use a cached response if it exists.
+A separate network request is made for that resource and the result is cached for future requests.
+
+This strategy is used when the most up to date version of a resource is not necessary for an application.
+
+| Route pattern                                     | Description          |
+| ------------------------------------------------- | -------------------- |
+| `/`                                               | The application root |
+| `/.\\.js$`                                        | JavaScript files     |
+| `/\/media\/catalog.*\.(?:png|gif|jpg|jpeg|svg)$/` | Catalog image files  |
```
Comment (supernova-at) [inline pwa-devdocs/src/technologies/basic-concepts/client-side-caching/index.md]: Not sure if Gatsby will treat this differently, but this table isn't showing up correctly:

Screen Shot 2019-04-23 at 10 25 51 AM

File: pwa-devdocs/src/technologies/basic-concepts/client-side-caching/index.md [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+
+You do not need to use Workbox to define service worker behavior, but
+Workbox makes this task easier by removing boilerplate code that is always used when working with service workers.
+
+Venia uses the following [caching strategies][] with its service worker:
+
+#### [Stale-while-revalidate][]
+
+The stale-while-revalidate strategy tells the service worker to use a cached response if it exists.
+A separate network request is made for that resource and the result is cached for future requests.
```
Comment (supernova-at) [inline pwa-devdocs/src/technologies/basic-concepts/client-side-caching/index.md]:
```suggestion
A separate network request is made for that resource and the cache is updated for future requests.
```

File: pwa-devdocs/src/technologies/basic-concepts/client-side-caching/index.md (hunk 6 above)
Comment (jcalcaben) [reply pwa-devdocs/src/technologies/basic-concepts/client-side-caching/index.md]: Looks like GitHub's markdown processor isn't a fan of the `|` inside the regex.
Jekyll's markdown processor seems to be able to build it correctly though.

**Pull request merged**